sycl/source/detail/plugin.hpp

```diff
@@ -110,12 +110,49 @@ class plugin {
   void *getLibraryHandle() const { return MLibraryHandle; }
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
+  // return the index os PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    auto It = std::find(PiPlatforms.begin(), PiPlatforms.end(), Platform);
+    if (It != PiPlatforms.end()) {
+      return It - PiPlatforms.begin();
+    } else {
+      PiPlatforms.push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms.size() - 1;
+    }
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
+    int PlatformId = getPlatformId(Platform);
+    if (PlatformId == 0)
+      return 0;
+    else
+      return LastDeviceIds[PlatformId - 1];
+  }
+  // set the id of the last device for the given platform
+  void setLastDeviceId(RT::PiPlatform Platform, int Id) {
+    int PlatformId = getPlatformId(Platform);
+    LastDeviceIds[PlatformId] = Id;
+  }
```

Comment:
I don't understand why we need all this complexity. Why can't we just add a method `getNewDeviceId` which would `return ++LastDeviceId`?

Reply:
We need to have LastDeviceId per platform.
The reason is that users can call platform::get_devices() multiple times for a given platform.
We can't simply keep incrementing the device id numbers because each invocation of platform::get_devices() will return different device ids.

Reply:
But we would reset to 0 in the beginning of `get_devices()`

Reply:
Correct. device::get_devices() will reset to zeros.
I was talking about the case where the users can call platform::get_devices() twice and but he did not call device::get_devices() in between.
It is a valid SYCL program, and we should return the same device ids for a given platform.

Reply:
I see. Instead of repeating the numbering, we should perform it once and remember devices in the cache, and their respectful IDs in the pi_device.

Reply:
Right. I agree, that would be ideal.
However, as we discussed before, it would require significant restructuring of our codes.
We are already doing caching of `sycl::device` at a later stage, but in our current implementation we redundantly call PI plugin calls to collect all devices first and then filter out, which still needs to decide the device id before we look up the `sycl::device` cache.

Reply:
We need to make our codes better. Let's do the proper device caching and then this code wouldn't be needed.

Reply:
To avoid calling PI plugins redundantly for every time `get_devices()` is called, we need to combine all the codes in device.cpp/h, device_impl.cpp/h, platform.cpp/h, platform_impl.cpp/h.
And possibly we need to make changes to our SYCL API spec, too.
We already discussed this before, and I can't decide this significant refactoring by myself.

@romanovvlad @vladimirlaz ,
Do you have any comments on Sergey's request?

Reply:
> To avoid calling PI plugins redundantly for every time `get_devices()` is called, we need to combine all the codes in device.cpp/h, device_impl.cpp/h, platform.cpp/h, platform_impl.cpp/h.

I disagree to this. There is certainly a way to implement right caching without combining all the code.

> And possibly we need to make changes to our SYCL API spec, too.

We would absolutely not extend SYCL language to do proper caching.

Reply:
It would be great if you elaborate "There is certainly a way to implement right caching without combining all the code."
I can certainly consider your suggestion if we can avoid restructuring the codes significantly.